packages/client/ui/react-ui/src/providers/CrossmintWalletProvider.tsx

```diff
@@ -0,0 +1,82 @@
+import { useCrossmint } from "@/hooks";
+import { kMaxLength } from "buffer";
+import { ReactNode, createContext, useEffect, useMemo, useState } from "react";
+
+import { EVMSmartWallet, SmartWalletError, SmartWalletSDK } from "@crossmint/client-sdk-smart-wallet";
+import { Crossmint } from "@crossmint/common-sdk-base";
+
+export type CrossmintWalletConfig = {
+    type: "evm-smart-wallet";
+    defaultChain: "polygon-amoy" | "base-sepolia" | "optimism-sepolia" | "arbitrum-sepolia";
+    createOnLogin: "all-users" | "off";
+};
+
+type WalletStatus = "not-loaded" | "in-progress" | "loaded" | "loading-error";
+
+type ValidWalletState =
+    | { status: "not-loaded" | "in-progress" }
+    | { status: "loaded"; wallet: EVMSmartWallet }
+    | { status: "loading-error"; error: SmartWalletError };
+
+function deriveErrorState(error: unknown): { status: "loading-error"; error: SmartWalletError } {
+    if (error instanceof SmartWalletError) {
+        return { status: "loading-error", error };
+    }
+
+    const message = error instanceof Error ? error.message : String(error);
+    const stack = error instanceof Error ? error.stack : undefined;
+    return { status: "loading-error", error: new SmartWalletError(`Unknown Wallet Error: ${message}`, stack) };
+}
+
+function shouldGetOrCreateWallet(status: WalletStatus, jwt: string | undefined): jwt is string {
+    return jwt != null && !(status === "in-progress" || status === "loaded");
+}
+
+type WalletContext = {
+    status: WalletStatus;
+    wallet?: EVMSmartWallet;
+    error?: SmartWalletError;
+    getOrCreateWallet: () => Promise<void>;
+};
+
+export const WalletContext = createContext<WalletContext>({
+    status: "not-loaded",
+    getOrCreateWallet: async () => {},
+});
+
+export function CrossmintWalletProvider({ children, config }: { config: CrossmintWalletConfig; children: ReactNode }) {
+    const { crossmint } = useCrossmint("CrossmintWalletProvider must be used within CrossmintProvider");
+    const [state, setState] = useState<ValidWalletState>({ status: "not-loaded" });
```

Comment:
i'm wondering if we should default this to "in-progress" so the initial render will be loading

Reply:
discussed in person